refl1d/probe/probe.py

```diff
@@ -1417,19 +1440,32 @@ def scattering_factors(self, material, density):
 
     scattering_factors.__doc__ = Probe.scattering_factors.__doc__
 
-    def oversample(self, n=20, seed=1):
+    def _apply_oversamplings(self):
+        calc_Q_parts = [self.Q]
+        if self.oversampling is not None:
+            calc_Q_parts.append(self._get_normal_oversampling_points(self.oversampling, self.oversampling_seed))
+        for region in self.oversampled_regions:
+            calc_Q_parts.append(np.linspace(region.Q_start, region.Q_end, region.n))
+        calc_Q = np.hstack(calc_Q_parts)
+        self.calc_Qo = np.sort(calc_Q)
+
+    def _get_normal_oversampling_points(self, n, seed):
         rng = numpy.random.RandomState(seed=seed)
         extra = rng.normal(self.Q, self.dQ, size=(n - 1, len(self.Q)))
-        calc_Q = np.hstack((self.Q, extra.flatten()))
-        self.calc_Qo = np.sort(calc_Q)
+        return extra
+
+    def oversample(self, n=20, seed=1):
+        self.oversampling = n
+        self.oversampling_seed = seed
+        self._apply_oversamplings()
 
     oversample.__doc__ = Probe.oversample.__doc__
 
     def critical_edge(self, substrate=None, surface=None, n=51, delta=0.25):
         Q_c = self.Q_c(substrate, surface)
-        extra = np.linspace(Q_c * (1 - delta), Q_c * (1 + delta), n)
-        calc_Q = np.hstack((self.Q, extra, 0))
-        self.calc_Qo = np.sort(calc_Q)
+        region = OversampledRegion(Q_start=Q_c * (1 - delta), Q_end=Q_c * (1 + delta), n=n)
+        self.oversampled_regions.append(region)
+        self._apply_oversamplings()
```

Comment:
Calling twice, for example in a jupyter notebook when you are exploring the choice of delta, will create a doubly oversampled part. I suggest keeping the critical region oversampling separate from the additional oversampled regions.

Reply:
Throw an error if there is already an oversampling specified, and perhaps warn that the method is deprecated.